### PR TITLE
[Mempool] Origination/Expiry Time validations

### PR DESCRIPTION
**Pull request by emmanueldenloye**
This PR addresses these items from the chainweb validations document/meetings:

* prop_tx_ttl_(validate/newblock)

a) tx origination time < block time,   b) tx origination time + tx TTL > block time  c) tx TTL <= maxTTL

In summary, the relationship between a transaction's origination and expiry time in in relation to the block time (here "block" is referring to the putative block that a transaction is included in) must satisfy the above set of inequalities.

### CODE REVIEW

File: src/Chainweb/Mempool/InMem.hs [hunk 1]
```diff
@@ -241,6 +255,12 @@ insertInMem cfg lock runCheck txs0 = do
     sizeOK tx = getSize tx <= maxSize
     maxRecent = _inmemMaxRecentItems cfg
     hasher = txHasher txcfg
+    ttlCheck tx =
+      case txMetadata txcfg tx of
+        TransactionMetadata (Time (TimeSpan creationTime)) (Time (TimeSpan expiryTime)) ->
+            creationTime >= 0 && expiryTime >= 0 && creationTime < expiryTime
+
+    preGossipCheck' = txPreGossipCheck txcfg
```
Comment (gregorycollins) [inline src/Chainweb/Mempool/InMem.hs]: I think we can delete this now that we have a pre-insert check in InMem?

File: src/Chainweb/Mempool/InMem.hs [hunk 2]
```diff
@@ -228,11 +230,23 @@ insertInMem cfg lock runCheck txs0 = do
                         else return
     preInsertFilterReal txs = do
         let chk = _inmemPreInsertCheck cfg
-        let out1 = V.map sizeOK txs
-        out2 <- chk txs
-        let oks = V.zipWith (&&) out1 out2
+
```
Comment (gregorycollins) [inline src/Chainweb/Mempool/InMem.hs]: Most of the new code in this function should probably just live inside `_inmemPreInsertCheck`

File: src/Chainweb/Mempool/InMem.hs [hunk 3]
```diff
@@ -228,11 +230,23 @@ insertInMem cfg lock runCheck txs0 = do
...
         return $! V.map fst $ V.filter snd $ V.zip txs oks
+      where
+         validators = [
+           sizeOK
+           , ttlCheck
+           , preGossipCheck'
+           ]
+         composeValidators xs ys zs =
+           foldl'
+           (\b validator -> V.zipWith (&&) b (validator <$> xs))
```
Comment (gregorycollins) [inline src/Chainweb/Mempool/InMem.hs]: Put a bang on `b`, use `<$!>`

File: src/Chainweb/Mempool/InMem.hs [hunk 4]
```diff
@@ -241,6 +255,12 @@ insertInMem cfg lock runCheck txs0 = do
     sizeOK tx = getSize tx <= maxSize
     maxRecent = _inmemMaxRecentItems cfg
     hasher = txHasher txcfg
+    ttlCheck tx =
+      case txMetadata txcfg tx of
+        TransactionMetadata (Time (TimeSpan creationTime)) (Time (TimeSpan expiryTime)) ->
+            creationTime >= 0 && expiryTime >= 0 && creationTime < expiryTime
```
Comment (gregorycollins) [inline src/Chainweb/Mempool/InMem.hs]: This is the wrong property -- should be

```
creationTime < now && expiryTime <= now
```

You'll have to get the current time at the top of the check

File: src/Chainweb/Miner/Coordinator.hs [hunk 5]
```diff
@@ -140,13 +140,14 @@ newWork choice miner pact c = do
             -- Fetch a Pact Transaction payload. This is an expensive call
             -- that shouldn't be repeated.
             --
-            payload <- _pactNewBlock pact miner p
+            creationTime <- getCurrentTimeIntegral
+
```
Comment (gregorycollins) [inline src/Chainweb/Miner/Coordinator.hs]: Kill blank line

File: src/Chainweb/Pact/PactService.hs [hunk 6]
```diff
@@ -263,7 +264,7 @@ serviceRequests memPoolAccess reqQ = do
                 go
             NewBlockMsg NewBlockReq {..} -> do
                 tryOne "execNewBlock" _newResultVar $
-                    execNewBlock memPoolAccess _newBlockHeader _newMiner
+                    execNewBlock memPoolAccess _newBlockHeader _newMiner _newCreationTime
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: Wrap long lines

File: src/Chainweb/Pact/PactService.hs [hunk 7]
```diff
@@ -420,18 +421,19 @@ _liftCPErr = either internalError' return
 validateChainwebTxsPreBlock
     :: PactDbEnv'
     -> Checkpointer
+    -> BlockCreationTime
     -> BlockHeight
     -> BlockHash
     -> Vector ChainwebTransaction
     -> IO (Vector Bool)
-validateChainwebTxsPreBlock dbEnv cp bh hash txs = do
+validateChainwebTxsPreBlock dbEnv cp (BlockCreationTime blockOriginationTime) bh hash txs = do
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: Wrap long lines

File: src/Chainweb/Pact/PactService.hs [hunk 8]
```diff
@@ -445,8 +447,11 @@ validateChainwebTxsPreBlock dbEnv cp bh hash txs = do
         let pactHash = view P.cmdHash tx
         mb <- _cpLookupProcessedTx cp pactHash
         if mb == Nothing
-        then checkAccount tx
+        -- prop_tx_ttl_newblock_validate
+        then (&&) <$> checkAccount tx <*> return (k tx)
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: Rather than making this single validator more complicated, keep it the same as before and compose single-purpose validations using a combinator

File: src/Chainweb/Pact/PactService.hs [hunk 9]
```diff
@@ -637,10 +643,18 @@ playOneBlock
 playOneBlock currHeader plData pdbenv = do
     miner <- decodeStrictOrThrow' (_minerData $ _payloadDataMiner plData)
     trans <- liftIO $ transactionsFromPayload plData
-    !results <- go miner trans
+
+    -- prop_tx_ttl_validateblock_validate
+    let validTrans = if isGenesisBlock then trans
+          else V.filter k trans
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: You can't filter transactions out of the block here. The block payload contains those transactions -- if any transaction in the block fails to validate then block validation should fail for the whole block. This is the reason why we've tried to be so thorough with the validations at newBlock time

Comment (gregorycollins) [reply src/Chainweb/Pact/PactService.hs]: Also -- transactions at the genesis block have to be valid according to the rules, they aren't special. If they are ill-formed then we need to create well-formed genesis txs, not skip validation here

File: src/Chainweb/Pact/Service/Types.hs [hunk 10]
```diff
@@ -71,6 +71,7 @@ type PactExMVar t = MVar (Either PactException t)
 data NewBlockReq = NewBlockReq
     { _newBlockHeader :: BlockHeader
     , _newMiner :: Miner
+    , _newCreationTime :: BlockCreationTime
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Service/Types.hs]: `StrictData` is on here right? I don't know why we turn it on, because the reviewer is always going to ask "is there a bang on here"

File: test/Chainweb/Test/Pact/ChainData.hs [hunk 11]
```diff
@@ -275,31 +283,17 @@ data TestException = TestException
...
-      Just (String pub) = HM.lookup "public" kp
-      Just (String priv) = HM.lookup "secret" kp
-      mkKeyBS = decodeKey . encodeUtf8
-  return $ ApiKeyPair (PrivBS $ mkKeyBS priv) (Just $ PubBS $ mkKeyBS pub) Nothing (Just ED25519)
-
-decodeKey :: ByteString -> ByteString
-decodeKey = fst . B16.decode
+  where
+    -- tx origination times needed to be unique to ensure that the corresponding
+    -- tx hashes are also unique.
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/ChainData.hs]: Just vary the nonces?

File: test/Chainweb/Test/Pact/PactInProcApi.hs [hunk 12]
```diff
@@ -21,10 +22,12 @@ import Control.Concurrent.Async
 import Control.Concurrent.MVar.Strict
 import Control.Concurrent.STM
 import Control.Exception
+import Control.Lens hiding ((.=))
 import Control.Monad (when)
 
 import Data.Aeson (object, (.=))
 import qualified Data.ByteString.Lazy as BL
+-- import Data.List (foldl')
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/PactInProcApi.hs]: remove cruft

File: src/Chainweb/Mempool/InMem.hs (hunk 1 above)
Comment (emmanueldenloye) [reply src/Chainweb/Mempool/InMem.hs]: Done.

File: src/Chainweb/Mempool/InMem.hs (hunk 4 above)
Comment (emmanueldenloye) [reply src/Chainweb/Mempool/InMem.hs]: done.

File: src/Chainweb/Mempool/InMem.hs (hunk 2 above)
Comment (emmanueldenloye) [reply src/Chainweb/Mempool/InMem.hs]: I cleaned up the code with this comment. The checks have been moved in a particular way. You'll see what I mean when you see the changes.

File: src/Chainweb/Mempool/InMem.hs (hunk 3 above)
Comment (emmanueldenloye) [reply src/Chainweb/Mempool/InMem.hs]: Done.

File: src/Chainweb/Mempool/InMem.hs (hunk 4 above)
Comment (emmanueldenloye) [reply src/Chainweb/Mempool/InMem.hs]: :open_mouth: oops... I've made the necessary change.

File: src/Chainweb/Miner/Coordinator.hs (hunk 5 above)
Comment (emmanueldenloye) [reply src/Chainweb/Miner/Coordinator.hs]: Done.

File: src/Chainweb/Pact/Service/Types.hs (hunk 10 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/Service/Types.hs]: StrictData isn't on in this module at all.

File: src/Chainweb/Pact/PactService.hs (hunk 8 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/PactService.hs]: done.

File: src/Chainweb/Pact/PactService.hs [hunk 13]
```diff
@@ -439,14 +441,19 @@ validateChainwebTxsPreBlock dbEnv cp bh hash txs = do
       m <- readCoinAccount dbEnv sender
       case m of
         Nothing -> return True
-        Just (T2 b _g) -> return $! b >= limitInCoin
+        Just (T2 b _g) -> do
+          let validations = and $ map ($ tx) validators
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: Prefer `all` here:

```
    Just (T2 b _g) ->
        let validations = (const (b >= limitInCoin)) : validators
        in return $! all ($ tx) validations
```

File: src/Chainweb/Pact/PactService.hs [hunk 14]
```diff
@@ -439,14 +441,19 @@ validateChainwebTxsPreBlock dbEnv cp bh hash txs = do
...
     checkOne tx = do
         let pactHash = view P.cmdHash tx
         mb <- _cpLookupProcessedTx cp pactHash
         if mb == Nothing
-        then checkAccount tx
+        -- prop_tx_ttl_newblock_validate
+        then checkAccount [k] tx
         else return False
+      where
+        k tx' = timingsCheck blockOriginationTime (payloadObj <$> tx')
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: Need to guard this w/ `blockHeight /= 0` also.

File: src/Chainweb/Pact/PactService.hs [hunk 15]
```diff
@@ -637,10 +648,31 @@ playOneBlock
 playOneBlock currHeader plData pdbenv = do
     miner <- decodeStrictOrThrow' (_minerData $ _payloadDataMiner plData)
     trans <- liftIO $ transactionsFromPayload plData
+
+    mapM_ checkTx trans
     !results <- go miner trans
     psStateValidated .= Just currHeader
     return $! toPayloadWithOutputs miner results
   where
+    checkTTL = if isGenesisBlock then const $ const True else timingsCheck
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: Add a TODO entry to remind ourselves that we need to dupe check in here. actually we should just do exactly the same newblock checks and delete all the duplicated code, but let's fix that later

File: src/Chainweb/Pact/PactService.hs [hunk 16]
```diff
@@ -803,6 +833,10 @@ applyPactCmd isGenesis dbEnv cmdIn miner mcache dl = do
         then applyGenesisCmd logger dbEnv pd spv cmd
         else applyCmd logger dbEnv miner (_psGasModel psEnv) pd spv cmd mcache
 
+    -- liftIO $ putStrLn "command result"
```
Comment (gregorycollins) [inline src/Chainweb/Pact/PactService.hs]: clean up

File: src/Chainweb/Pact/RestAPI/Server.hs [hunk 17]
```diff
@@ -37,7 +37,7 @@ import Control.Monad.Trans.Maybe
 import Data.Aeson as Aeson
 import qualified Data.ByteString.Lazy as LBS
 import qualified Data.ByteString.Base64.URL.Lazy as Base64
-import qualified Data.ByteString.Short as SB
+-- import qualified Data.ByteString.Short as SB
```
Comment (gregorycollins) [inline src/Chainweb/Pact/RestAPI/Server.hs]: clean up

File: src/Chainweb/Pact/RestAPI/Server.hs [hunk 18]
```diff
@@ -415,5 +415,6 @@ validateCommand :: Command Text -> Either String ChainwebTransaction
 validateCommand cmdText = let
   cmdBS = encodeUtf8 <$> cmdText
   in case verifyCommand cmdBS of
-  ProcSucc cmd -> return $! (\bs -> PayloadWithText (SB.toShort bs) (_cmdPayload cmd)) <$> cmdBS
+  ProcSucc cmd -> return $! mkPayloadWithText <$> cmd
+  -- (\bs -> PayloadWithText (SB.toShort bs) (_cmdPayload cmd)) <$> cmdBS
```
Comment (gregorycollins) [inline src/Chainweb/Pact/RestAPI/Server.hs]: clean up

File: src/Chainweb/Pact/Service/BlockValidation.hs [hunk 19]
```diff
@@ -36,13 +36,14 @@ import Chainweb.Payload
 import Chainweb.Transaction
 
 
-newBlock :: Miner -> BlockHeader -> TQueue RequestMsg ->
+newBlock :: Miner -> BlockHeader -> BlockCreationTime -> TQueue RequestMsg ->
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Service/BlockValidation.hs]: Reformat according to style guide 

```
newBlock
    :: Miner
    -> BlockHeader
    -> ...
```

File: src/Chainweb/Pact/Service/Types.hs [hunk 20]
```diff
@@ -46,6 +46,7 @@ data PactException
   = BlockValidationFailure Text
   | PactInternalError Text
   | NoBlockValidatedYet
+  | TransactionValidationException [(PactHash, Text)]
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Service/Types.hs]: We don't need to collect them all (I changed the routine to throw on first error) so let's make this

```
| TransactionValidationException !PactHash !Text
```

File: src/Chainweb/Pact/TransactionExec.hs [hunk 21]
```diff
@@ -182,7 +182,7 @@ applyGenesisCmd logger dbEnv pd spv cmd = do
 
     resultE <- catchesPactError $! runPayload cmdEnv initState cmd []
     fmap (`T2` mempty) $! case resultE of
-      Left e ->
+      Left e -> do
```
Comment (gregorycollins) [inline src/Chainweb/Pact/TransactionExec.hs]: Revert this change

File: test/Chainweb/Test/Pact/ChainData.hs [hunk 22]
```diff
@@ -112,12 +105,27 @@ chainDataTest t =
...
+    -> BlockHeight
+    -> BlockHash
+    -> IO (V.Vector ChainwebTransaction)
+getTestBlock t txOrigTime _validate _bh _hash = do
     akp0 <- stockKey "sender00"
     kp0 <- mkKeyPairs [akp0]
-    nonce <- (<> t) . T.pack . show @(Time Int) <$> getCurrentTimeIntegral
-    mkTestExecTransactions "sender00" "0" kp0 nonce 10000 0.00000000001 3600 0 tx
+    let nonce = (<> t) . T.pack . show @(Time Integer) $ txOrigTime
+    txs <- mkTestExecTransactions "sender00" "0" kp0 nonce 10000 0.00000000001 3600 (toTxCreationTime txOrigTime) tx
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/ChainData.hs]: wrap long line

File: test/Chainweb/Test/Pact/ChainData.hs [hunk 23]
```diff
@@ -275,31 +283,17 @@ data TestException = TestException
 
 instance Exception TestException
 
-testMemPoolAccess :: T.Text -> MemPoolAccess
-testMemPoolAccess t = MemPoolAccess
-    { mpaGetBlock = \_ _ _ _ -> getTestBlock t
+testMemPoolAccess :: T.Text -> IO (Time Integer) -> MemPoolAccess
+testMemPoolAccess t iotime = MemPoolAccess
+    { mpaGetBlock = \validate bh hash _parentHeader -> do
+        time <- meinhack bh <$> iotime
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/ChainData.hs]: `meinhack` lol change this

File: test/Chainweb/Test/Pact/PactReplay.hs [hunk 24]
```diff
@@ -98,56 +104,78 @@ onRestart pdb bhdb r = do
...
-    ksData :: Text -> Value
-    ksData idx = object [("k" <> idx) .= object [ "keys" .= ([] :: [Text]), "pred" .= String ">=" ]]
-    getTestBlock validate _bHeight _bHash bHeader = do
-        let Nonce nonce = _blockNonce bHeader
-            moduleStr = defModule (T.pack $ show nonce)
-            d = Just $ ksData (T.pack $ show nonce)
-        let txs = V.fromList $ [PactTransaction moduleStr d]
-        outtxs <- goldenTestTransactions txs
-        oks <- validate _bHeight _bHash outtxs
+    meinhack :: BlockHeight -> Time Integer -> Time Integer
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/PactReplay.hs]: same here

File: test/Chainweb/Test/Pact/PactReplay.hs [hunk 25]
```diff
@@ -98,56 +104,78 @@ onRestart pdb bhdb r = do
...
+                           , show oks
+                           ]
         return outtxs
-
-dupegenMemPoolAccess :: MemPoolAccess
-dupegenMemPoolAccess  = MemPoolAccess
-    { mpaGetBlock = getTestBlock
+      where
+        ksData :: Text -> Value
+        ksData idx = object [("k" <> idx) .= object [ "keys" .= ([] :: [Text]), "pred" .= String ">=" ]]
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/PactReplay.hs]: wrap long lines

File: test/Chainweb/Test/Pact/Utils.hs [hunk 26]
```diff
@@ -193,12 +206,13 @@ mkTestExecTransactions sender cid ks nonce0 gas gasrate ttl ct txs = do
       case verifyCommand cmd of
         ProcSucc t ->
           let
-            r = fmap (k t) $ SB.toShort <$> cmd
+            -- r = fmap (k t) $ SB.toShort <$> cmd
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/Utils.hs]: clean up

File: test/Chainweb/Test/Pact/Utils.hs [hunk 27]
```diff
@@ -193,12 +206,13 @@ mkTestExecTransactions sender cid ks nonce0 gas gasrate ttl ct txs = do
...
+            r = mkPayloadWithText <$> t
             -- order matters for these tests
           in return $ (succ n, Vector.snoc acc r)
         ProcFail e -> throwM $ userError e
 
-    k t bs = PayloadWithText bs (_cmdPayload t)
+    -- k t bs = PayloadWithText bs (_cmdPayload t)
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/Utils.hs]: tidy

File: test/Chainweb/Test/Pact/Utils.hs [hunk 28]
```diff
@@ -238,10 +252,11 @@ mkTestContTransaction sender cid ks nonce gas rate step pid rollback proof ttl c
 
     cmd <- mkCommand ks pm nonce msg
     case verifyCommand cmd of
-      ProcSucc t -> return $ Vector.singleton $ fmap (k t) (SB.toShort <$> cmd)
+      -- ProcSucc t -> return $ Vector.singleton $ fmap (k t) (SB.toShort <$> cmd)
+      ProcSucc t -> return $ Vector.singleton $ mkPayloadWithText <$> t
       ProcFail e -> throwM $ userError e
   where
-    k t bs = PayloadWithText bs (_cmdPayload t)
+    -- k t bs = PayloadWithText bs (_cmdPayload t)
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/Utils.hs]: tidy

File: test/Chainweb/Test/Pact/Utils.hs [hunk 29]
```diff
@@ -446,3 +460,32 @@ withPactCtxSQLite v cutDB bhdbIO pdbIO f =
...
+  , "keys" .= ks
+  ]
+
+stockKeyFile :: ByteString
+stockKeyFile = $(embedFile "pact/genesis/testnet/keys.yaml")
+
+-- | Convenient access to predefined testnet sender accounts
+stockKey :: Text -> IO ApiKeyPair
+stockKey s = do
+  let Right (Y.Object o) = Y.decodeEither' stockKeyFile
```
Comment (gregorycollins) [inline test/Chainweb/Test/Pact/Utils.hs]: we indent four characters round these parts pardner

File: tools/ea/Ea.hs [hunk 30]
```diff
@@ -31,11 +31,13 @@ module Ea ( main ) where
...
 import BasePrelude
 
+import Control.Lens (set)
+
 import Data.Aeson (ToJSON)
 import Data.Aeson.Encode.Pretty
 import Data.ByteString (ByteString)
 import qualified Data.ByteString.Lazy as BL
-import qualified Data.ByteString.Short as SB
+-- import qualified Data.ByteString.Short as SB
```
Comment (gregorycollins) [inline tools/ea/Ea.hs]: tidy

File: tools/ea/Ea.hs [hunk 31]
```diff
@@ -55,11 +57,15 @@ import Chainweb.Graph
 import Chainweb.Logger (genericLogger)
 import Chainweb.Miner.Pact (noMiner)
 import Chainweb.Pact.PactService
+-- import Chainweb.Payload
```
Comment (gregorycollins) [inline tools/ea/Ea.hs]: tidy

File: tools/ea/Ea.hs [hunk 32]
```diff
@@ -119,7 +125,9 @@ genPayloadModule v tag txFiles =
                 procCmd = verifyCommand cmdBS
             case procCmd of
                 f@ProcFail{} -> fail (show f)
-                ProcSucc c -> fixupPayload cmdBS c
+                ProcSucc c -> do
+                  let t = toTxCreationTime (Time (TimeSpan 0))
```
Comment (gregorycollins) [inline tools/ea/Ea.hs]: 4 space indent on do

File: tools/standalone/Standalone/Utils.hs [hunk 33]
```diff
@@ -10,7 +10,7 @@
 
 module Standalone.Utils where
 
-import Control.Concurrent
+-- import Control.Concurrent
```
Comment (gregorycollins) [inline tools/standalone/Standalone/Utils.hs]: tidy

File: tools/standalone/Standalone/Utils.hs [hunk 34]
```diff
@@ -19,7 +19,7 @@ import Control.Lens hiding ((.=))
 import Data.Aeson
 import Data.ByteString (ByteString)
 import qualified Data.ByteString.Base16 as B16
-import qualified Data.ByteString.Short as SB
+-- import qualified Data.ByteString.Short as SB
```
Comment (gregorycollins) [inline tools/standalone/Standalone/Utils.hs]: tidy

File: tools/standalone/Standalone/Utils.hs [hunk 35]
```diff
@@ -124,26 +121,11 @@ defaultMemPoolAccess cid blocksize  = MemPoolAccess
               ks <- testKeyPairs
               cmd <- mkCommand ks pm nonce msg
               case verifyCommand cmd of
-                ProcSucc t -> return $ fmap (k t) (SB.toShort <$> cmd)
+                ProcSucc t -> return $ mkPayloadWithText <$> t
+                -- fmap (k t) (SB.toShort <$> cmd)
```
Comment (gregorycollins) [inline tools/standalone/Standalone/Utils.hs]: :put_litter_in_its_place:

File: tools/standalone/Standalone/Utils.hs [hunk 36]
```diff
@@ -124,26 +121,11 @@ defaultMemPoolAccess cid blocksize  = MemPoolAccess
...
-          (Right !pdo) -> return pdo
-          Left e -> throwM e
-  , _pactNewBlock = \m h -> do
-      mv <- newBlock m h q
-      r <- takeMVar mv
-      case r of
-          (Right !pdo) -> return pdo
-          Left e -> throwM e
-  }
+          -- k t bs = PayloadWithText bs (_cmdPayload t)
```
Comment (gregorycollins) [inline tools/standalone/Standalone/Utils.hs]: :put_litter_in_its_place:

File: tools/standalone/Standalone/Utils.hs [hunk 37]
```diff
@@ -245,10 +227,10 @@ mkExecTransactions cid ks nonce0 gas gasrate ttl ct txs = do
       let nonce = T.append nonce0 (T.pack $ show nn)
       cmd <- mkCommand ks pm nonce msg
       case verifyCommand cmd of
-        ProcSucc t -> return $ fmap (k t) (SB.toShort <$> cmd)
+        ProcSucc t -> return $! mkPayloadWithText <$> t
         ProcFail e -> throwM $ userError e
 
-    k t bs = PayloadWithText bs (_cmdPayload t)
+    -- k t bs = PayloadWithText bs (_cmdPayload t)
```
Comment (gregorycollins) [inline tools/standalone/Standalone/Utils.hs]: :put_litter_in_its_place:

File: src/Chainweb/Pact/PactService.hs (hunk 15 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/PactService.hs]: done

File: src/Chainweb/Pact/PactService.hs (hunk 14 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/PactService.hs]: done

File: src/Chainweb/Pact/PactService.hs (hunk 13 above)
Comment (emmanueldenloye) [reply src/Chainweb/Pact/PactService.hs]: done
